File: sorting-searching/sorting-algorithms/External-merge-sort/external-merge-sort.js
const fs = require("fs");

const CHUNK_SIZE = 4;

function createSortedRuns(inputFile) {
  const data = fs.readFileSync(inputFile, "utf8")
    .trim()
    .split(/\s+/)
    .map(Number);

  let runIndex = 0;

  for (let i = 0; i < data.length; i += CHUNK_SIZE) {
    const chunk = data
      .slice(i, i + CHUNK_SIZE)
      .sort((a, b) => a - b);

    fs.writeFileSync(
      `run${runIndex}.txt`,
      chunk.join(" ")
    );

    runIndex++;
  }

  return runIndex;
}

function mergeArrays(left, right) {
  const result = [];

  let i = 0;
  let j = 0;

  while (i < left.length && j < right.length) {
    if (left[i] <= right[j]) {
      result.push(left[i++]);
    } else {
      result.push(right[j++]);
    }
  }

  while (i < left.length) {
    result.push(left[i++]);
  }

  while (j < right.length) {
    result.push(right[j++]);
  }

  return result;
}

function mergeRuns(totalRuns, outputFile) {
  let merged = [];

  for (let i = 0; i < totalRuns; i++) {
    const runData = fs.readFileSync(
      `run${i}.txt`,
      "utf8"
    )
      .trim()
      .split(/\s+/)
      .map(Number);

    merged = mergeArrays(merged, runData);
  }

  fs.writeFileSync(outputFile, merged.join(" "));
}

fs.writeFileSync(
  "input.txt",
  "8 3 7 4 9 2 6 1"
);

const totalRuns = createSortedRuns("input.txt");

mergeRuns(totalRuns, "sorted_output.txt");

console.log("External Merge Sort completed.");
console.log("Check sorted_output.txt");

File: sorting-searching/sorting-algorithms/External-merge-sort/external-merge-sort.py
import heapq

CHUNK_SIZE = 4

def create_sorted_runs(input_file):
    with open(input_file, "r") as file:
        data = list(map(int, file.read().split()))

    run_index = 0

    for i in range(0, len(data), CHUNK_SIZE):
        chunk = data[i:i + CHUNK_SIZE]
        chunk.sort()

        with open(f"run{run_index}.txt", "w") as run_file:
            run_file.write(" ".join(map(str, chunk)))

        run_index += 1

    return run_index


def merge_runs(total_runs, output_file):
    run_arrays = []

    for i in range(total_runs):
        with open(f"run{i}.txt", "r") as file:
            run_arrays.append(
                list(map(int, file.read().split()))
            )

    merged = list(heapq.merge(*run_arrays))

    with open(output_file, "w") as output:
        output.write(" ".join(map(str, merged)))


with open("input.txt", "w") as file:
    file.write("8 3 7 4 9 2 6 1")

total_runs = create_sorted_runs("input.txt")

merge_runs(total_runs, "sorted_output.txt")

print("External Merge Sort completed.")
print("Check sorted_output.txt")
